src/SchematicHQ.Client/Datastream/Client.cs

```diff
@@ -858,7 +858,161 @@ private async Task GetAllFlagsAsync(CancellationToken cancellationToken)
+            return false;
+        }
+
+        if (eventBody.Company == null || eventBody.Company.Count == 0)
+        {
+            _logger.Error("No keys provided for company lookup");
+            return false;
+        }
+
+        // Get company from cache
+        var company = GetCompanyFromCache(eventBody.Company);
+        if (company == null)
+        {
+            return false;
+        }
+
+        // Create a deep copy of the company to avoid modifying the cached object directly
+        var companyCopy = DeepCopyCompany(company);
+        if (companyCopy == null)
+        {
+            _logger.Error("Failed to create deep copy of company");
+            return false;
+        }
+
+        // Update the metric value if it matches the event
+        bool metricUpdated = false;
+        foreach (var metric in companyCopy.Metrics)
+        {
+            if (metric != null && metric.EventSubtype == eventBody.Event)
+            {
+                int quantity = eventBody.Quantity ?? 0;
+                metric.Value += quantity;
+                metricUpdated = true;
+            }
+        }
+
+        if (!metricUpdated)
+        {
+            _logger.Debug($"No matching metric found for event {eventBody.Event}");
+            return false;
+        }
+
+        // Cache the updated company for all keys
+        bool cacheSuccess = true;
+        foreach (var key in companyCopy.Keys)
+        {
+            try
+            {
+                var cacheKey = ResourceKeyToCacheKey<Company>(CacheKeyPrefixCompany, key.Key, key.Value);
+                _companyCache.Set(cacheKey, companyCopy);
+            }
+            catch (Exception ex)
+            {
+                _logger.Warn($"Failed to cache company metric for key '{key.Key}': {ex.Message}");
+                cacheSuccess = false;
+            }
+        }
+
+        return cacheSuccess;
+    }
+
+    /// <summary>
+    /// Creates a complete deep copy of a Company object and all its nested fields.
+    /// This ensures that modifying the returned company won't affect the original cached object.
+    /// </summary>
+    /// <param name="company">The company to copy</param>
+    /// <returns>A new independent copy of the company</returns>
+    private Company? DeepCopyCompany(Company? company)
+    {
+        if (company == null)
+        {
+            return null;
+        }
+
+        // Create a new company instance
+        var companyCopy = new Company
+        {
+            Id = company.Id,
+            AccountId = company.AccountId,
+            EnvironmentId = company.EnvironmentId,
+            BasePlanId = company.BasePlanId,
+            BillingProductIds = new List<string>(company.BillingProductIds),
+            CrmProductIds = new List<string>(company.CrmProductIds),
+            PlanIds = new List<string>(company.PlanIds),
+            Subscription = company.Subscription, // Shallow copy as we don't modify it
```

Comment:
Why is a shallow copy ok here, given we are potentially going to write this whole object back ito the cache after copy?